balance_strategy.go

```diff
@@ -127,3 +155,899 @@ func balanceStrategyHashValue(vv ...string) uint32 {
+}
+
+// The assignment should improve the overall balance of the partition assignments to consumers.
+func assignPartition(partition topicPartitionAssignment, sortedCurrentSubscriptions []string, currentAssignment map[string][]topicPartitionAssignment, consumer2AllPotentialPartitions map[string][]topicPartitionAssignment, currentPartitionConsumer map[topicPartitionAssignment]string) []string {
+	for _, memberID := range sortedCurrentSubscriptions {
+		if memberAssignmentsIncludeTopicPartition(consumer2AllPotentialPartitions[memberID], partition) {
+			currentAssignment[memberID] = append(currentAssignment[memberID], partition)
+			currentPartitionConsumer[partition] = memberID
+			break
+		}
+	}
+	return sortMemberIDsByPartitionAssignments(currentAssignment)
+}
+
+// Deserialize topic partition assignment data to aid with creation of a sticky assignment.
+func deserializeTopicPartitionAssignment(userDataBytes []byte) (StickyAssignorUserData, error) {
+	userDataV1 := &StickyAssignorUserDataV1{}
+	if err := decode(userDataBytes, userDataV1); err != nil {
+		userDataV0 := &StickyAssignorUserDataV0{}
+		if err := decode(userDataBytes, userDataV0); err != nil {
+			return nil, err
+		}
+		return userDataV0, nil
+	}
+	return userDataV1, nil
+}
+
+// filterAssignedPartitions returns a map of consumer group members to their list of previously-assigned topic partitions, limited
+// to those topic partitions currently reported by the Kafka cluster.
+func filterAssignedPartitions(currentAssignment map[string][]topicPartitionAssignment, partition2AllPotentialConsumers map[topicPartitionAssignment][]string) map[string][]topicPartitionAssignment {
+	assignments := deepCopyAssignment(currentAssignment)
+	for memberID, partitions := range assignments {
+		// perform in-place filtering
+		i := 0
+		for _, partition := range partitions {
+			if _, exists := partition2AllPotentialConsumers[partition]; exists {
+				partitions[i] = partition
+				i++
+			}
+		}
+		assignments[memberID] = partitions[:i]
+	}
+	return assignments
+}
+
+func removeTopicPartitionFromMemberAssignments(assignments []topicPartitionAssignment, topic topicPartitionAssignment) []topicPartitionAssignment {
+	for i, assignment := range assignments {
+		if assignment == topic {
+			return append(assignments[:i], assignments[i+1:]...)
+		}
+	}
+	return assignments
+}
+
+func memberAssignmentsIncludeTopicPartition(assignments []topicPartitionAssignment, topic topicPartitionAssignment) bool {
+	for _, assignment := range assignments {
+		if assignment == topic {
+			return true
+		}
+	}
+	return false
+}
+
+func sortPartitions(currentAssignment map[string][]topicPartitionAssignment, partitionsWithADifferentPreviousAssignment map[topicPartitionAssignment]consumerGenerationPair, isFreshAssignment bool, partition2AllPotentialConsumers map[topicPartitionAssignment][]string, consumer2AllPotentialPartitions map[string][]topicPartitionAssignment) []topicPartitionAssignment {
```

Comment:
After staring at `performReassignments` for a while, I'm pretty sure this entire function is useless. This entire thing can be deleted and replaced with just returning the keys of partitions2AllPotentialConsumers and the tests pass.

Really, the only benefit of this function is if all group subscriptions are identical. In that case, the return ordering is actually meaningful. `performReassignments` will start from partitions that are on overloaded members, and thus it may find balance sooner. Even the loss of that ordering can be largely mitigated by saving if all subscriptions are identical and then using that in `isBalanced`: if everything is identical, the max imbalance should be one, and the remainder of the function can be skipped.

If there's any disparity of consuming, the return is pretty useless: it sorts by partitions with the least amount of potential consumers. That does not help reassigning.

Reply:
For further context, even if all subscriptions are identical, the only aid is in _maybe_ less looping. `performReassignments` _always_ visits all partitions, left to right. If a partition is on an overloaded member, it gets moved. If a move happens, _all_ "sorted" partitions are revisited left to right. After the first round of moves, sortedPartitions is no longer sorted, yet reassigning still happens in the original order.

I'm honestly not quite sure what sorting here buys at all.

Reply:
When all can consume the same, it may help in the case of incoming joins with prior plans
```
A -> 1
B-> 2, 3, 4
C -> 5, 6, 7, 8, 9
```
Sorted would say `9, 8, 7, 4, 6, 3, 5, 2, 1` in order of move, which would result in
```
A -> 1, 9, 8
B -> 2, 3, 4
C -> 5, 6, 7
```
and be most sticky,
whereas no sorting would be `1, 2, 3, 4, 5, 6, 7, 8, 9` order of move, resulting in
```
A -> 1, 2, 5
B -> 3, 4, 6
C -> 7, 8, 9
```
being destructive to B when it shouldn't have. I'll see if there's a test case similar to this.

I'm still mixed on how beneficial sorting is for non-equal consuming and will try to think of a scenario there.

Reply:
This test does that:
```
                {
                        name: "Three consumers with a disparity coming into harmony.",
                        args: args{
                                members: map[string]ConsumerGroupMemberMetadata{
                                        "c1": ConsumerGroupMemberMetadata{
                                                Topics:   []string{"t1"},
                                                UserData: encodeSubscriberPlanWithGeneration(t, map[string][]int32{"t1": []int32{0}}, 1),
                                        },
                                        "c2": ConsumerGroupMemberMetadata{
                                                Topics:   []string{"t1"},
                                                UserData: encodeSubscriberPlanWithGeneration(t, map[string][]int32{"t1": []int32{1, 2, 3}}, 1),
                                        },   
                                        "c3": ConsumerGroupMemberMetadata{
                                                Topics:   []string{"t1"},
                                                UserData: encodeSubscriberPlanWithGeneration(t, map[string][]int32{"t1": []int32{4, 5, 6, 7, 8}}, 1),
                                        },   
                                },   
                                topics: map[string][]int32{
                                        "t1": []int32{0, 1, 2, 3, 4, 5, 6, 7, 8},
                                },
                        },   
                },
```
And printing the plan with and without sortPartitions shows the problem. Unfortunately, the test code doesn't catch the lack of stickiness without the sorting:
```
=== RUN   Test_stickyBalanceStrategy_Plan/Three_consumers_with_a_disparity_coming_into_harmony.
c1 => map[t1:[0 2 6]]
c2 => map[t1:[1 3 8]]
c3 => map[t1:[7 5 4]]
```
That should fail, since c2 should not have changed.

Reply:
As an example of why the second sorting is useless:
```go
                {
                        // A -> 1, can take all
                        // B -> 2, 3, 4
                        // C -> 5, 6, 7, 8, 9
                        //
                        // Bad would be stealing a partition from B:
                        // A -> 1, 2, 5
                        // B -> 3, 4
                        // C -> 6, 7, 8, 9
                        //
                        // Ideal:
                        // A -> 1, 5, 6
                        // B -> 2, 3, 4
                        // C -> 7, 8, 9
                        name: "Bigly disbalancy.",
                        args: args{
                                members: map[string]ConsumerGroupMemberMetadata{
                                        "A": ConsumerGroupMemberMetadata{
                                                Topics: []string{"1", "2", "3", "4", "5", "6", "7", "8", "9"},
                                                UserData: encodeSubscriberPlanWithGeneration(t,
                                                        map[string][]int32{
                                                                "1": []int32{0},
                                                        }, 1),
                                        },
                                        "B": ConsumerGroupMemberMetadata{
                                                Topics: []string{"2", "3", "4"},
                                                UserData: encodeSubscriberPlanWithGeneration(t,
                                                        map[string][]int32{
                                                                "2": []int32{0},
                                                                "3": []int32{0},
                                                                "4": []int32{0},
                                                        }, 1),
                                        },
                                        "C": ConsumerGroupMemberMetadata{
                                                Topics: []string{"5", "6", "7", "8", "9"},
                                                UserData: encodeSubscriberPlanWithGeneration(t,
                                                        map[string][]int32{
                                                                "5": []int32{0},
                                                                "6": []int32{0},
                                                                "7": []int32{0},
                                                                "8": []int32{0},
                                                                "9": []int32{0},
                                                        }, 1),
                                        },   
                                },
                                topics: map[string][]int32{
                                        "1": []int32{0},
                                        "2": []int32{0},
                                        "3": []int32{0},
                                        "4": []int32{0},
                                        "5": []int32{0},
                                        "6": []int32{0},
                                        "7": []int32{0},
                                        "8": []int32{0},
                                        "9": []int32{0},
                                },
                        },
```
That causes a test failure. I think the heap sort is really the only beneficial sort: we should always try to move partitions off of loaded members. I'll try to think of another example.

Reply:
Here's another test that fails with either heap strategy:
```
                {
                        // A -> 1, [in all]
                        // B -> 2, [in 2, 3, 4]
                        // D -> 3, 4, 5, 6, 7, 8, 9
                        //
                        // A -> 1, 9, 8, 7
                        // B -> 2,
                        // D -> 3, 4, 5, 6
                        //
                        // A -> 1, 9, 8, 7
                        // B -> 2, 3
                        // D -> 4, 5, 6
                        //
                        // Ideal:
                        // A -> 1, 9, 8
                        // B -> 2, 3, 4
                        // C -> 5, 6, 7
                        name: "Bigly disbalancy.",
                        args: args{
                                members: map[string]ConsumerGroupMemberMetadata{
                                        "A": ConsumerGroupMemberMetadata{
                                                Topics: []string{"1", "2", "3", "4", "5", "6", "7", "8", "9"},
                                                UserData: encodeSubscriberPlanWithGeneration(t,
                                                        map[string][]int32{
                                                                "1": []int32{0},
                                                        }, 1),
                                        },
                                        "B": ConsumerGroupMemberMetadata{
                                                Topics: []string{"2", "3", "4"},
                                                UserData: encodeSubscriberPlanWithGeneration(t,
                                                        map[string][]int32{
                                                                "2": []int32{0},
                                                        }, 1),
                                        },
                                        "C": ConsumerGroupMemberMetadata{
                                                Topics: []string{"3", "4", "5", "6", "7", "8", "9"},
                                                UserData: encodeSubscriberPlanWithGeneration(t,
                                                        map[string][]int32{
                                                                "3": []int32{0},
                                                                "4": []int32{0},
                                                                "5": []int32{0},
                                                                "6": []int32{0},
                                                                "7": []int32{0},
                                                                "8": []int32{0},
                                                                "9": []int32{0},
                                                        }, 1),
                                        },
                                },
                                topics: map[string][]int32{
                                        "1": []int32{0},
                                        "2": []int32{0},
                                        "3": []int32{0},
                                        "4": []int32{0},
                                        "5": []int32{0},
                                        "6": []int32{0},
                                        "7": []int32{0},
                                        "8": []int32{0},
                                        "9": []int32{0},
                                },
                        },
                },
```
I think maybe a better strategy would be to always try reassigning _from_ the consumer with the least partitions, and iterating over all partitions that could be assigned to it. I'll try that out.

Reply:
Alright, I have a scheme that solves all three cases above:

1) always sort by partitions-on-most-loaded-member
2) In `performReassignments`, start from the least loaded member (with sortedCurrentSubscriptions)
3) look for the _first_ partition in `reassignablePartitions` that it can take,
4) if the current consumer of that partition is not itself, and the current consumer is more loaded by more than 1, take the partition

I think this also largely obviates the need for `prevAssignment`.

Reply:
After more playing around, I think something like this is ideal:
```go
411         for {
 412                 modified = false
 413                 // reassign all reassignable partitions (starting from the partition with least potential consumers and if needed)
 414                 // until the full list is processed or a balance is achieved
 415                 if isBalanced(currentAssignment, sortedCurrentSubscriptions, consumer2AllPotentialPartitions) {
 416                         break
 417                 }
 418                 for _, leastLoaded := range sortedCurrentSubscriptions {
 419                         potentialPartitions := consumer2AllPotentialPartitions[leastLoaded]
 420                         myConsumptions := len(currentAssignment[leastLoaded])
 421                         var mostConsumptions int
 422                         var mostPartition topicPartitionAssignment
 423                         for _, partition := range potentialPartitions {
 424                                 currentConsumer := currentPartitionConsumer[partition]
 425                                 if currentConsumer == leastLoaded {
 426                                         continue
 427                                 }
 428 
 429                                 currentConsumptions := len(currentAssignment[currentConsumer])
 430                                 if currentConsumptions > myConsumptions {
 431                                         if currentConsumptions > mostConsumptions {
 432                                                 mostConsumptions = currentConsumptions
 433                                                 mostPartition = partition
 434                                         }
 435                                 }
 436                         }
 437                         if mostConsumptions > 0 {
 438                                 sortedCurrentSubscriptions = s.reassignPartitionToNewConsumer(mostPartition, currentAssignment, sortedCurrentSubscriptions, currentPartitionConsumer, con     sumer2AllPotentialPartitions)   
 439                                 reassignmentPerformed = true
 440                                 modified = true
 441                                 break
 442                         }
 443                 }
 444                 if !modified {
 445                         return reassignmentPerformed
 446                 }       
```

I don't expect you to use this, but I do think it provides optimal balancing as well as stickiness. Modulo some bugs, that's the strategy I plan to use in my WIP client.

Reply:
A more complicated one:
```
                {
                        // Start:
                        // A -> 1 2
                        // B -> 3 4
                        // C -> 5
                        // D -> a b c d e
                        // E ->
                        //
                        // A takes all,
                        // B takes 1-5
                        // C takes 3-5
                        // D, E take a-e
                        //
                        // Ideal:
                        // A -> 1 e
                        // B -> 2 3
                        // C -> 4 5
                        // D -> a b
                        // E -> c d
                        name: "Complicated",
                        args: args{
                                members: map[string]ConsumerGroupMemberMetadata{
                                        "A": ConsumerGroupMemberMetadata{
                                                Topics: []string{"1", "2", "3", "4", "5", "a", "b", "c", "d", "e"},
                                                UserData: encodeSubscriberPlanWithGeneration(t,
                                                        map[string][]int32{
                                                                "1": []int32{0},
                                                                "2": []int32{0},
                                                        }, 1),
                                        },
                                        "B": ConsumerGroupMemberMetadata{
                                                Topics: []string{"1", "2", "3", "4", "5"},
                                                UserData: encodeSubscriberPlanWithGeneration(t,
                                                        map[string][]int32{
                                                                "3": []int32{0},
                                                                "4": []int32{0},
                                                        }, 1),
                                        },
                                        "C": ConsumerGroupMemberMetadata{
                                                Topics: []string{"3", "4", "5"},
                                                UserData: encodeSubscriberPlanWithGeneration(t,
                                                        map[string][]int32{
                                                               "4": []int32{0},
                                                        }, 1),
                                        },
                                        "D": ConsumerGroupMemberMetadata{
                                                Topics: []string{"a", "b", "c", "d", "e"},
                                                UserData: encodeSubscriberPlanWithGeneration(t,
                                                        map[string][]int32{
                                                                "a": []int32{0},
                                                                "b": []int32{0},
                                                                "c": []int32{0},
                                                                "d": []int32{0},
                                                                "e": []int32{0},
                                                        }, 1),
                                        },
                                        "E": ConsumerGroupMemberMetadata{
                                                Topics: []string{"a", "b", "c", "d", "e"},
                                                UserData: encodeSubscriberPlanWithGeneration(t,
                                                        map[string][]int32{}, 1),
                                        },
                                },
                                topics: map[string][]int32{
                                        "1": []int32{0},
                                        "2": []int32{0},
                                        "3": []int32{0},
                                        "4": []int32{0},
                                        "5": []int32{0},
                                        "6": []int32{0},
                                        "7": []int32{0},
                                        "8": []int32{0},
                                        "9": []int32{0},
                                        "a": []int32{0},
                                        "b": []int32{0},
                                        "c": []int32{0},
                                        "d": []int32{0},
                                        "e": []int32{0},
                                },
                        },
```

Reply:
FWIW, I figured out a way to solve all of these edge cases and implemented it here: https://github.com/twmb/kgo/tree/master/internal/sticky

The key idea is representing the members and partitions they want as a graph with edges, and then using A* search to steal appropriately.

This implementation is faster, more accurate (for the balanced-metric, and I think optimally accurate), uses less garbage, and is cleaner than the Java code.

Reply:
After optimizing, the implementation in the above link is >14x faster.